### PR TITLE
Enabling Helix perf test runs for CoreCLR

### PR DESCRIPTION
**Pull request by dsgouda**
Added targets to pack, upload and trigger perf test runs in CoreCLR
- Created new targets file for perf related tasks, added call to this target in runtest.proj and runtest.cmd
- Created AfterTargets for build of individual tests which copies over the project.json files for individual tests into the bin directories, this allows for assembly list generation for the corresponding tests when uploading as a workitem
- Updated version numbers of xunit performance packages to correspond with latest version supported by latest build tools version

@rahku @brianrob @jhendrixMSFT @MattGal @brianrob @lt72 

P.S.: I will fix the build tools version numbers once review is completed

### CODE REVIEW

File: tests/src/dir.targets [hunk 1]
```diff
@@ -76,6 +76,11 @@
     <MakeDir Condition="'$(CLRTestKind)' == 'RunOnly'" ContinueOnError="false" Directories="$(OutputPath)" />
   </Target>
   
+  <Target Name="CopyJsonProjectFiles" AfterTargets="Build">  
+     <MSBuild Projects="@(ProjectReference)" Condition="Exists('%(RootDir)%(Directory)%(Filename)%(Extension)')"/>
```
Comment (rahku) [inline tests/src/dir.targets]: why is the project being built again?

Comment (rahku) [inline tests/src/dir.targets]: Is it possible to only perform this for perf tests. It seems this is happening for all the tests which is overkill as perf tests are only a small portion.

File: tests/helix.targets [hunk 2]
```diff
@@ -7,12 +7,12 @@
 
   <ItemGroup>
     <!-- TODO This should come from some package -->
-    <CoreRootFiles Include="$(CORE_ROOT)\crossgen.exe" /> 
-    <CoreRootFiles Include="$(CORE_ROOT)\ildasm.exe" /> 
-    <CoreRootFiles Include="$(CORE_ROOT)\ilasm.exe" /> 
+    <CoreRootFiles Include="$(CORE_ROOT)\crossgen.exe" />
```
Comment (rahku) [inline tests/helix.targets]: why are all these unnecessary changes showing up? Can you get rid of all these unnecessary changes throughout your PR.

File: tests/runtest.proj [hunk 3]
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="src\dir.props" />
-  <Import Project="$(ToolsDir)Build.Post.targets" Condition="Exists('$(ToolsDir)Build.Post.targets') AND '$(BuildWrappers)' == 'true'" /> 
+  <Import Project="$(ToolsDir)Build.Post.targets" Condition="Exists('$(ToolsDir)Build.Post.targets') AND '$(BuildWrappers)' == 'true'" />
+  <Import Project="helixperftasks.targets" />
```
Comment (rahku) [inline tests/runtest.proj]: import only for perf runs?

File: tests/runtest.proj [hunk 4]
```diff
@@ -338,32 +340,32 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
...
+
     <!-- Restore all wrapper projects' project.jsons in one pass for perf & to avoid concurrency problems -->
     <Exec Command="$(DnuRestoreCommand) &quot;$(XunitWrapperGeneratedCSDirBase.TrimEnd('\\'))&quot;"
           StandardOutputImportance="Low" />
-    
+
     <Message Importance="High" Text="[$([System.DateTime]::Now.ToString('HH:mm:ss.ff'))] Restoring all packages...Done." />
 
-    <MSBuild Projects="$(MSBuildProjectFile)" Targets="BuildXunitWrapper" Properties="_CMDDIR=%(TestDirectories.Identity)" />
+    <MSBuild Projects="$(MSBuildProjectFile)" Targets="BuildXunitWrapper" Properties="_CMDDIR=%(TestDirectories.Identity)" Condition="'$(Performance)'!='true'" />
```
Comment (rahku) [inline tests/runtest.proj]: why is this condition here? You should not running this whole target (createAllWrappers) for perf run

File: tests/runtest.proj [hunk 5]
```diff
@@ -412,16 +414,29 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
...
 
+  <Target Name="RunPerfTests">
+    <Message Text="Executing steps for perf tests" Importance="High" Condition="'$(Performance)'=='true'"/>
+    <!-- generate project.json for runtime dependency -->
+    <MSBuild Projects="$(MSBuildProjectFile)"
+             Targets="CreateTestRuntimeJsonFile"
+             Condition=" '$(BuildWrappers)'=='true' OR '$(Performance)'=='true' " />
+
+    <!-- Package the product -->
+    <MSBuild Projects="$(MSBuildProjectFile)" Targets="CreateTestOverlay" />
```
Comment (rahku) [inline tests/runtest.proj]: this task is already executed via runtest.cmd. Can you reuse that instead of calling here?

File: tests/runtest.cmd [hunk 6]
```diff
@@ -191,6 +192,21 @@ if errorlevel 1 (
     exit /b 1
 )
 
+:RunPerfTests
```
Comment (rahku) [inline tests/runtest.cmd]: It seems like this would run even for functional runs. Is that what we want?

File: tests/helixperftasks.targets [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      <OutDir>$([System.IO.Path]::GetDirectoryName($(CurrPerfAssembly)))/</OutDir>
+      <TestArchiveDir>$(BaseOutputPath)\testStagingDir\</TestArchiveDir>
+      <TestProjectName>$([System.IO.Path]::GetFileNameWithoutExtension($(CurrPerfAssembly)))</TestProjectName>
+      <TestRuntimeDependenciesJson>$(SourceDir)Common\test_runtime\project.json</TestRuntimeDependenciesJson>
+      <TestRuntimeProjectLockJson>$(SourceDir)Common\test_runtime\project.lock.json</TestRuntimeProjectLockJson>
+      <ProjectLockJson>$(OutDir)project.lock.json</ProjectLockJson>
+    </PropertyGroup>
+
+    <!-- Restore the runtime dependencies -->
+    <Exec Command="$(DnuRestoreCommand) &quot;$(TestRuntimeDependenciesJson)&quot;"
```
Comment (rahku) [inline tests/helixperftasks.targets]: seems like this should happen only once and not for every perf test

**Comment by rahku** [comment]
Looks good except for above review comments. Thanks for doing this.

File: tests/runtest.proj (hunk 5 above)
Comment (dsgouda) [inline tests/runtest.proj]: If I understand correctly that happens in :ResolveDependencies: This triggers the "CloudBuild" targets upon completion which in turn tries to find the zipped Perf tests and upload them (which it can't find since I find and zip the perf tests later in helixperftest.targets)
I wanted to remove this coupling of targets hence the isolated call

File: tests/helix.targets (hunk 2 above)
Comment (dsgouda) [inline tests/helix.targets]: Its my sublime text going berserk with whitespaces, will try to fix and repost these changes

File: tests/src/dir.targets (hunk 1 above)
Comment (dsgouda) [inline tests/src/dir.targets]: I couldn't find an easy way to figure out if a test is a Perf test while building a project itself, I can add the condition if we can distinguish the perf tests when building them

Comment (dsgouda) [inline tests/src/dir.targets]: We are building the project to get hold of the $(ProjectJson) property from the csproj tests if I understand correctly. @lt72 can you confirm this

File: tests/helixperftasks.targets (hunk 7 above)
Comment (dsgouda) [inline tests/helixperftasks.targets]: We are copying over the project.json file corresponding to each project (which varies for each test project in CoreCLR), hence running the restore on each project.json corresponding to the test before generating an assembly list for it

Comment (rahku) [inline tests/helixperftasks.targets]: that restore should have already happened as part of buildtest.cmd All that would be remaining is restore for runtime. That only needs to happen once.

**Comment by rahku** [comment]
Looks good now...except for the one outstanding comment above as to why the project is being built again. Also it would be nice if I can run the perf tests locally on my machine as well. Does this happen already as part of functional test runs?

**Comment by dsgouda** [comment]
Thanks Rahul. Unfortunately there is no way to run Perf/functional tests locally, that would require further work. Ideally, it would be better if we can piggyback perf local test runs on the implementation for functional tests. Could you take a look into the effort required for it?

**Comment by rahku** [comment]
So are the perf tests being excluded from functional runs? How does that exclusion happen...do you know? I have seen tests in jit\performance directory being executed as part of functional runs.

**Comment by dsgouda** [comment]
The Perf tests are not excluded from functional tests as of now. By piggybacking I meant implementing something like what we have for CoreFX where we can build and run any individual test by copying over all the binaries and dependencies in the test directory.
Eg: we can do
msbuild src\System.Runtime.Serialization.Xml\tests\System.Runtime.Serialization.Xml.Tests.csproj /t:rebuildandtest /p:Performance=true /p:Configuration=Windows_Release
to build and execute a single test (we can run functional tests without the /p:Performance=true param)
@mattgal may have more context here

File: tests/src/dir.targets (hunk 1 above)
Comment (dsgouda) [inline tests/src/dir.targets]: Talked to @lt72, since this is an aftertargets on build we are not exactly doing the build again, just adding a target to perform the copy after each test is built.

**Comment by rahku** [comment]
OK so we can run perf tests locally (as it happens as part of functional runs) is all that I was looking for.

**Comment by dsgouda** [comment]
Yes, since it happens as a part of functional tests and we can run functional tests locally, the answer would be yes.

File: tests/src/dir.targets (hunk 1 above)
Comment (rahku) [inline tests/src/dir.targets]: As long as it is just copy should be fine. But it does not look like to me. Just to be clear I am asking about below line in the target.
<MSBuild Projects="@(ProjectReference)" Condition="Exists('%(RootDir)%(Directory)%(Filename)%(Extension)')"/>
The next line does make sense as it seems to be doing the copy. But I am no msbuild expert and could be wrong.

Comment (rahku) [inline tests/src/dir.targets]: Text got removed in above comment. This is the line I am asking about. 
MSBuild Projects="@(ProjectReference)" Condition="Exists('%(RootDir)%(

Comment (dsgouda) [inline tests/src/dir.targets]: After digging further I realized that project.json is not set for a bunch of projects and hence the Copy operation does not work on them. For all the perf tests however the ProjectJson property is set across all csproj files and hence the project.json files can be copied over (This discrepancy is rather confusing) Will update the PR with changes
